### DISCUSSION TITLE
giving subjects a unique ID in order to give them a unique set of images

### DISCUSSION THREAD

**Discussion by csavasegal** [jspsych/jsPsych#2255] [Q&A]
Hello,
I'm interested in doing a true randomization of my files. Specifically, I have 100 images and I want to present a unique set to every single one of my participants without having it so that one of the images gets seen more than the others (I want an even spread). Given that I won't have more than 50 participants, just randomly selecting images will not work. 
It seems like the best way to do that is to come up with a unique set of images for each of my participants prior to people signing up. So if I have 50 participants, I will have an external file that has 50 entries with 50 unique lists of images and then I can read from this external file each time the experiment is run.

However, I do not know how to ensure that each participant that signs up on prolific and is given my jsPsych experiment is given a unique number from 1 to 50 as opposed to multiple subjects being given the same one. **Is it possible to keep track so that the first person that signs up is 1, then 2 and so on? Or do I need to use a random list generator and generator a number between 1 and 50 for all of my subjects?** 

This seems to be a common issue (true randomization) in psychological research so I hope someone else has come up with a solution. Thank you!

**Comment by jodeleeuw** [comment]
> Is it possible to keep track so that the first person that signs up is 1, then 2 and so on?

Yes, this is possible, but it requires using some kind of server-side storage like a database. How are you planning to host your experiment?

**Reply by csavasegal** [reply]
I'm currently hosting on pavlovia. Would that work?

**Reply by becky-gilbert** [reply]
In case it's helpful: one way to think about this is as version counterbalancing rather than randomization. In other words, you could think of your 50 unique stimuli lists as 50 separate 'versions' of your experiment, and your aim is to assign one participant per version. I don't have experience with Pavlovia but I'm guessing they have a system for version counterbalancing, so you could look into that and see if it works for this situation.

**Reply by csavasegal** [reply]
Thanks for the tip! 
Unfortunately, there doesn't seem to be an existent way to counterbalance on pavlovia. I have found the suggestion to use this on many forums: https://moryscarter.com/vespr/pavlovia.php and am wondering how I could define the 'participant' value to the value given by this php. I realize that you may not know this, but figured it I would put it here in case someone has tried this same implementation. 

I have tried to use the server phps that are here: https://github.com/vassar-cogscilab based on what I've seen as responses on other posts, but they do not seem to work for me. **Is there a way to modify these codes to use pavlovia instead of a personal server?** 
Thanks very much for the help!

**Reply by RebeccaHirst** [reply]
Hello! 

Heads up that I am not familiar with jsPsych here so I'm speaking from a the perspective of a Pavlovia from PsychoPy perspective and relying on your jsPsych expertise to build the bridge here! 

OK so in pavlovia the easiest way to pass variables to your experiment is via query string. So if your experiment URL looked like this 
https://run.pavlovia.org/USERNAME/EXPERIMENT_NAME/

You could update this to be https://run.pavlovia.org/USERNAME/EXPERIMENT_NAME/?participant=1234&group=A
(here we are assigning a participant value and group value, what the morys carter app is doing is generating a sequentially increasing participant ID, that can then be used for participant assignment - which is probably what you already figured out). 

Now, here's the step I unfortunately don't know... inside a PsychoPy/PsychoJS experiment any values passed in via query string are accessible from within the experiment (and therefore can be used to assign a set of images) using `expInfo[PARAMETER]` so we could call `expInfo['participant']` or `expInfo['group']` - my question for Javascript peeps **is there a way to access values passed through query string from within a jsPsych experiment?** if so then you can use the morys carter app in exactly the same way (i.e. access that participant ID and then assign images based on that ID, as you suggest).

**Reply by csavasegal** [reply]
@RebeccaHirst Thanks so much! This is great. 
I was planning to update for others once I got it fully working, but I've been able to integrate the morys carter app successfully and then I use this line: `var participant = jsPsych.data.getURLVariable('participant');` 

My plan (haven't fully implemented properly yet) is to use this variable to reference a list of stimuli from a .js script and then use that to call the stimuli in a for loop in my plug-ins.

Thanks everyone and I hope this helps someone else in the future!

**Reply by becky-gilbert** [reply]
Thanks @RebeccaHirst! I don't know much about Pavlovia so really appreciate your input! In general, is there a Pavlovia support forum/email we could refer people to when we get Pavlovia-related questions?

Also, just to make sure that the answer to this question is super clear to any jsPsych users:
> my question for Javascript peeps is there a way to access values passed through query string from within a jsPsych experiment? 

Yes, the [`jsPsych.data.getURLVariable` function](https://www.jspsych.org/7.0/reference/jspsych-data/#jspsychdatageturlvariable) gets the value of a specific URL variable, and the [`jsPsych.data.urlVariables` function](https://www.jspsych.org/7.0/reference/jspsych-data/#jspsychdataurlvariables) will get all URL query parameters and store them in an object. So in the case of your example link: https://run.pavlovia.org/USERNAME/EXPERIMENT_NAME/?participant=1234&group=A
you can do this in the jsPsych experiment code to retrieve the values of the URL variables:
```js
var participant = jsPsych.data.getURLVariable('participant'); // "1234"
var group = jsPsych.data.getURLVariable('group'); // "A"
```
or this:
```js
var urls_vars = jsPsych.data.urlVariables();
var participant = urls_vars.participant; // "1234"
var group = urls_vars.group; // "A"
```

**Reply by RebeccaHirst** [reply]
Hi @becky-gilbert - yes absolutely - if any user has pavlovia specific Qs I would encourage to post on the help forum https://discourse.psychopy.org/ and categorise the question under "online experiments" or with the tag "pavlovia" - hope this helps!

It is also handy for me to know about accessing URL variables in jsPsych so thanks for that! (I must improve my familiarity!) 
Thanks!
Becca

**Reply by becky-gilbert** [reply]
Great, thanks Becca!
